### PR TITLE
Integrate RunGPT For LLMs

### PR DESCRIPTION
**Pull request by selmiss**
# Description

**RunGPT** is a wonderful open-source LLM model serving framework belong to JinaAI company. It is designed to simplify the deployment and management of large language models. LlamaIndex can do use this kind of model service to accelerate model deployment and enhance the user experience using llm service.

Integrate RunGPT through using class LLM at llama_index/llms/rungpt.py

Add required Jupter notebooks at docs/examples/llm/rungpt.ipynb

Add required unit test at tests/llms/test_rungpt.py

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Install necessary packages such as run-gpt-torch, and run a simple command to start a service of large language model localhost. The detail tutorial could be found in docs/examples/llm/rungpt.ipynb.

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x]  I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: llama_index/llms/rungpt.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    ):
+        super().__init__(
+            model=model,
+            endpoint=endpoint,
+            temperature=temperature,
+            max_tokens=max_tokens,
+            context_window=context_window,
+            additional_kwargs=additional_kwargs or {},
+            callback_manager=callback_manager or CallbackManager([]),
+            base_url="http://" + endpoint,
```
Comment (logan-markewich) [inline llama_index/llms/rungpt.py:59]: nit: might be more robust to check if the endpoint starts with `http` before adding it?

File: llama_index/llms/rungpt.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        )
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        """LLM metadata."""
+        return LLMMetadata(
+            context_window=self.context_window,
+            num_output=DEFAULT_NUM_OUTPUTS,
```
Comment (logan-markewich) [inline llama_index/llms/rungpt.py:67]: this should be `self.max_tokens` I think?

File: llama_index/llms/rungpt.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
+from typing import Any, Optional, Sequence, Dict
+from pydantic import Field
+import json
+import requests
+import sseclient
```
[the remark this answers is not in the document]
Comment (logan-markewich) [reply llama_index/llms/rungpt.py:5]: Looks like it's also causing unit tests to fail currently

[the remark this answers is not in the document]
Comment (selmiss) [reply llama_index/llms/rungpt.py:5]: Yes, but what should we do to pass the unit test without adding packages in requirements.txt?

File: llama_index/llms/rungpt.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+        response_gpt = requests.post(
+            self.base_url + "/generate",
+            json=self._request_pack("complete", prompt, **kwargs),
+            stream=False,
+        ).json()
+
+        response_gpt = CompletionResponse(
+            text=response_gpt["choices"][0]["text"],
+            additional_kwargs=response_gpt["usage"],
```
[the remark this answers is not in the document]
Comment (selmiss) [reply llama_index/llms/rungpt.py:82]: Yes, `usage` contains the token counting information, and it is available for all the endpoint. I will add this right now.

File: llama_index/llms/rungpt.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (logan-markewich) [reply llama_index/llms/rungpt.py:5]: Check out other tests that rely on external packages 

Either they need to be skipped if not installed, or you need to mock a bunch of stuff 

https://github.com/jerryjliu/llama_index/blob/3c338ea59be0bc9b4b98bce3fdc6be895409852a/tests/llms/test_anthropic.py#L11

File: llama_index/llms/rungpt.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+                "messages": prompt,
+                "max_tokens": kwargs.pop("max_tokens", self.max_tokens),
+                "temperature": kwargs.pop("temperature", self.temperature),
+                "top_k": kwargs.pop("top_k", 50),
+                "top_p": kwargs.pop("top_p", 0.95),
+                "repetition_penalty": kwargs.pop("repetition_penalty", 1.2),
+                "do_sample": kwargs.pop("do_sample", False),
+                "echo": kwargs.pop("echo", True),
+                "n": kwargs.pop("n", 1),
+                "stop": kwargs.pop("stop", "."),
```
[the remark this answers is not in the document]
Comment (selmiss) [reply llama_index/llms/rungpt.py:279]: Yes, your point is correct. Setting a stop marker may cause the model to stop prematurely while generating text, resulting in less than complete or sensible text being generated. This is a trade-off that needs to be made, depending on the application scenario and requirements. In short, the appropriate stop marker settings need to be adjusted on a case-by-case basis to achieve the best possible generation results.

File: llama_index/llms/rungpt.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+                delta = item_dict["choices"][0]["text"]
+                text = text + delta
+                yield CompletionResponse(text=text, delta=delta, raw=item_dict)
+
+        return gen()
+
+    @llm_chat_callback()
+    def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
+        message_list = list()
+        for message in messages:
```
[the remark this answers is not in the document]
Comment (selmiss) [reply llama_index/llms/rungpt.py]: Two helper function `_message_wrapper` and `_message_unpacker` come for this.

File: llama_index/llms/rungpt.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (selmiss) [reply llama_index/llms/rungpt.py]: OK, I have fixed it in the latest commit.

**Comment by selmiss** [comment]
@logan-markewich 
Thank you for your feedback and suggestions. I appreciate your help in identifying the bug and guiding me towards a solution. The bug that was occurring during the `make lint` process has been resolved.

**Comment by logan-markewich** [comment]
Thanks @selmiss !! 💪🏻

**Pull request merged**